File: C/nombresDiosMT.c
/*
 *      nombresDios1_0.c
 *      
 *      Copyright 2009 josemi <josemi@darksun>
 *      
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; either version 2 of the License, or
 *      (at your option) any later version.
 *      
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *      
 *      You should have received a copy of the GNU General Public License
 *      along with this program; if not, write to the Free Software
 *      Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *      MA 02110-1301, USA.
 */

/* linux gcc nombresDiosMT.c -std=c99 -O3 -fopenmp -march=x86-64 -o nombresDiosMT */

#include <stdio.h>
#include <time.h>
#include <omp.h>

#define CAR  (13)
#define POS  (9)
#define MAXBLOCK  (3)
#define TRUE (1)
#define FALSE (0)

long long sum = 0;
long long sumOK = 0;
/* supongo long long 64 bits */

void element(int first_row)
{
		
	int cell[POS];
	int ended = FALSE;
	
	int l,k,j,e;

	/*
	int th_id;
	th_id = omp_get_thread_num();
    	printf("Hello World from thread %d\n", th_id);
	*/	

	for(l=0; l<POS-1;l++) cell[l]=0;
	cell[POS-1] = first_row;
		
		
		do {	
			sum++;
			
			int numOK = TRUE;
			for(k=0; (numOK) && (k<POS-MAXBLOCK);k++)
			{
				numOK = FALSE;
				for(e = k + 1; (!numOK) && (e < k+MAXBLOCK+1); e++)
				{
					
					if(cell[k]!=cell[e])
					{
						
						numOK = TRUE;
					}
				}
				
			}
				
			
			if(numOK)
			{
				sumOK++;
				/*
				StringBuffer a= new StringBuffer();
				for(int j=POS-1;j>=0;j--)
				{
					a.append((char)(65+cell[j]));
				}
				System.out.println(a);
				*/
				/*
				for(j=POS-1;j>=0;j--)
				{
					printf("%c",65+cell[j]);
				}
				printf("\n");
				*/
			}
			
			
			cell[0]++;
			int i=0;
			while(cell[i]>=CAR)
			{
				cell[i]=0;
				i++;
				if(i<POS-1)
				{
					cell[i]++;
				}
				else
				{
					ended = TRUE;
					break;
				}
			}

		} while(!ended);


}


int main(int argc, char** argv)
{

	printf("Version %ld, CAR %d, POS %d, MAXBLOCK %d\n",sizeof(int)*8,CAR,POS,MAXBLOCK);		

	omp_set_num_threads(CAR);		

	//clock_t begin = clock();
	//clock give as proccess time, for MP is way bigger than wall clock
	//This function give us the wall clock
	double begin = omp_get_wtime();	

	#pragma omp parallel for reduction(+:sum,sumOK)	
	for(int iTh = 0; iTh < CAR; iTh++)
	{
		element(iTh);
	}		

		
	//clock_t end = clock();
	double end = omp_get_wtime();	
	
	printf("Sum   %lld \n",sum);
		
	printf("SumOK %lld \n",sumOK);
		
	printf("Time: %f \n",((end-begin)*1000.0));

	return 0;
}

File: java/nombresDiosMT.java
/*
NamesOfGod

Copyright (C) 2016 Jose Miguel Garrido

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicInteger;

/**
 *
 * @author josemi
 */
public class nombresDiosMT {

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
	// 13, 9, 3
	final int CAR = 13;
	final int POS = 9;
	final int MAXBLOCK = 3;

        // somos atomicos!!!
        final AtomicLong sumMT = new AtomicLong(0);
        final AtomicLong sumOKMT = new AtomicLong(0);
        final AtomicInteger thNum = new AtomicInteger(0);

	System.out.println("Caracteres   "+CAR);
	System.out.println("Posiciones   "+POS);
	System.out.println("Repeticiones "+MAXBLOCK);

        long begin = System.currentTimeMillis();

        final Thread[] pool = new Thread[CAR];
        for (int iTh = 0; iTh < CAR; iTh++)
        {
            pool[iTh] = new Thread()
            {
                @Override
                public void run()
                {


                    // cada thread tiene
                    int cell[] = new int[POS];
                    boolean ended = false;
                    long sum = 0;
                    long sumOK = 0;

                    cell[POS-1]=thNum.getAndIncrement();
                    //System.out.println("Arrancado thread "+cell[POS-1]);

                    do
                    {
                        sum++;

                        boolean numOK = true;
                        for(int k=0; (numOK) && (k<POS-MAXBLOCK);k++)
                        {
                            numOK = false;
                            for(int e = k + 1; (!numOK) && (e < k+MAXBLOCK+1); e++)
                            {
                                //System.out.println("k: "+k+" e: "+e);
                                if(cell[k]!=cell[e])
                                {
                                    //System.out.println("Verdad");
                                    numOK = true;
                                }
                            }

                        }


                        if(numOK)
                        {
                            sumOK++;
                            /*
                            StringBuffer a= new StringBuffer();
                            for(int j=POS-1;j>=0;j--)
                            {
                                a.append((char)(65+cell[j]));
                            }
                            System.out.println(a);
                            */
                        }


                        cell[0]++;
                        int i=0;
                        while(cell[i]>=CAR)
                        {
                            cell[i]=0;
                            i++;
                            // ojo, ahora llegamos a POS-1, POS esta fijado para
                            // el thread
                            if(i<POS-1)
                            {
                                cell[i]++;
                            }
                            else
                            {
                                ended = true;
                                break;
                            }
                        }

                    } while(!ended);

                    // hemos acabado para este thread, sumamos
                    sumMT.getAndAdd(sum);
                    sumOKMT.getAndAdd(sumOK);

                } // de run
            }; // de thread

            // arrancamos los CAR Threads
            pool[iTh].start();

        }  // del for

        // esperamos a que acaben todos
        for (Thread t : pool)
        {
            try
            {
                t.join();
            }
            catch (InterruptedException e)
            {
                e.printStackTrace();
            }
        }


		long end = System.currentTimeMillis();

		System.out.println("MT Sum   "+sumMT);

		System.out.println("MT SumOK "+sumOKMT);

		System.out.println("MT Time: "+(end-begin));
	}


}
